[WebGPU+Swift] Eliminate 'unsafe' caused by QuerySet.CounterSampleBuffer

geoffreygaren opened the pull request:
#### 26fdd36cde06b4614dea7d8dbc47f4c8581168d9
<pre>
[WebGPU+Swift] Eliminate &apos;unsafe&apos; caused by QuerySet.CounterSampleBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=302055">https://bugs.webkit.org/show_bug.cgi?id=302055</a>
<a href="https://rdar.apple.com/164135057">rdar://164135057</a>

Reviewed by Mike Wyrzykowski.

Swift treats std::pair&lt;id, int&gt; as unsafe because

(a) there&apos;s a pointer in there

(b) Swift has no way of knowing whether the pointer&apos;s lifetime is guaranteed or
not -- because Swift models C++ copy and move constructors and copy and move
assignment operators as opaque functions.

Switched from std::pair to a custom struct so I could declare it SWIFT_ESCAPABLE.
I also took the opporutnity to give the data members more descriptive names than
&apos;first&apos; and &apos;second&apos;.

Removed a Range(uncheckedBounds:) construction because it&apos;s unsafe and
we can use literal range syntax intead.

Changed some nil short-circuiting to use the ?? operator, for style.

Fixed up some stray &apos;;&apos;, for style.

Canonical link: <a href="https://commits.webkit.org/302666@main">https://commits.webkit.org/302666@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e5b5dc31ce23d7f1778c27b60f2d394809c7190

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129843 "16 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2104 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40700 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137235 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81325 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/19329727-deaa-4880-a2b1-3641c2b6aee4) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2056 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1994 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/98913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e21603fb-30db-46eb-abab-2c9d9757d5fe) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132790 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1562 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116287 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79598 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f8bf385e-1f37-4433-9e85-6a90244b63cf) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80506 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109974 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34921 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139715 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1898 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/1773 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107420 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1943 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112634 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107300 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1532 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31118 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54699 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20253 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1971 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65340 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1785 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1820 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1894 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -837,27 +837,25 @@ extension WebGPU.CommandEncoder {
-        if m_device.ptr().enableEncoderTimestamps() || counterSampleBuffer != nil {
-            if counterSampleBuffer != nil {
+        if m_device.ptr().enableEncoderTimestamps() || counterSampleBuffer.buffer != nil {
+            if counterSampleBuffer.buffer != nil {
```
these should all definitely really be like ```guard let buffer = counterSampleBuffer.buffer else { // handle error }``` but idk if that’s too disruptive for this PR or not

geoffreygaren replied:
`guard` would be incorrect here. From [the documentation](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/statements/#Guard-Statement):

```
The else clause of a guard statement is required, and must either call a function
with the Never return type or transfer program control outside the guard statement’s
enclosing scope using one of the following statements:

return
break
continue
throw
```

rr-codes replied:
sorry ya I should’ve been more specific; if there is any error handling like early returning or throwing done then use guard let whatever = something , otherwise use if let whatever = something 

comparison to nil should almost never be seen in most Swift code

mwyrzykowski replied:
Seems outside the scope of this PR 👍

geoffreygaren replied:
For the first nil comparison, as far as I can tell there's no `if let` syntax that permits the use of `||`.

For the second nil comparison, I think I can do an `if let`.

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -1896,15 +1894,13 @@ extension WebGPU.CommandEncoder {
-            let counterSampleBufferOffset = unsafe querySet.counterSampleBufferWithOffset().second
-            guard let counterSampleBuffer else {
+            let counterSampleBuffer = querySet.counterSampleBufferWithOffset()
+            guard counterSampleBuffer.buffer != nil else {
```
guard let here too

geoffreygaren replied:
Unfortunately `guard let x else { }` syntax is only allowed when `x` is a local identifier. In this case, `x` is a property expression.

rr-codes replied:
sorry by this I meant `guard let buffer = counterSampleBuffer.buffer`

geoffreygaren replied:
That suggestion introduces the declaration of an unused variable, which is such a hazard to readability that it is usually considered an an error in Swift (although apparently permitted in this case for now), and is at least considered a warning in many other programming environments.

I don't think the need to remove comparison to nil is so urgent that it should justify coding patterns that are often considered warnings or errors.

rr-codes replied:
but it wouldn’t be unused? line 1902 would use it

geoffreygaren replied:
Yeah, that's fair, I can change that line to match.

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -1896,15 +1894,13 @@ extension WebGPU.CommandEncoder {
-                counterSampleBuffer,
-                range: Range(uncheckedBounds: (Int(firstQuery + counterSampleBufferOffset), Int(firstQuery + counterSampleBufferOffset + queryCount))),
+                counterSampleBuffer.buffer,
+                range: Range(uncheckedBounds: (Int(firstQuery + counterSampleBuffer.offset), Int(firstQuery + counterSampleBuffer.offset + queryCount))),
```
usually a Range is just made via like (x + y)..<(x + y + z)

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -2150,24 +2146,22 @@ extension WebGPU.CommandEncoder {
+        if m_device.ptr().enableEncoderTimestamps() || counterSampleBuffer.buffer != nil {
             let timestampWrites = wgpuGetComputePassDescriptorTimestampWrites(collection.span)?[0]
-            computePassDescriptor.sampleBufferAttachments[0].sampleBuffer = counterSampleBuffer != nil ? computePassDescriptor.sampleBufferAttachments[0].sampleBuffer : m_device.ptr().timestampsBuffer(m_commandBuffer, 2)
+            computePassDescriptor.sampleBufferAttachments[0].sampleBuffer = counterSampleBuffer.buffer != nil ? computePassDescriptor.sampleBufferAttachments[0].sampleBuffer : m_device.ptr().timestampsBuffer(m_commandBuffer, 2)
```
this condition can be simplified it looks like? can just be like

```
if counterSampleBuffer.buffer == nil {
  computePassDescriptor.sampleBufferAttachments[0].sampleBuffer = m_device.ptr().timestampsBuffer(m_commandBuffer, 2)
}
```

right?

geoffreygaren replied:
I'm going to change this to

`counterSampleBuffer.buffer ?? m_device.ptr().timestampsBuffer(m_commandBuffer, 2)`.

That's not quite the same thing! But it mimics the C++ code better:

`counterSampleBuffer.buffer ?: m_device->timestampsBuffer(m_commandBuffer, 2);`

So I think we had a transcription error at some point.

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -2150,24 +2146,22 @@ extension WebGPU.CommandEncoder {
 
-            if counterSampleBuffer != nil {
-                m_device.ptr().trackTimestampsBuffer(m_commandBuffer, counterSampleBuffer)
+            if counterSampleBuffer.buffer != nil {
```
if let

geoffreygaren replied:
Unfortunately you can only `if let` a local identifier, not an expression 😞 .

rr-codes replied:
sorry I meant if let buffer = counterSampleBuffer.buffer

mwyrzykowski commented on Source/WebGPU/WebGPU/QuerySet.h:
```diff
@@ -54,7 +54,10 @@ class Device;
 public:
-    using CounterSampleBuffer = std::pair<id<MTLCounterSampleBuffer>, uint32_t>;
+    struct CounterSampleBuffer {
+        id<MTLCounterSampleBuffer> buffer { nil }; // Safety: ARC retains this pointer
```
Is the comment necessary since all of WebGPU is ARC and extensively uses `id<T>` member variables?

mwyrzykowski replied:
or are we just documenting it due to the `SWIFT_ESCAPABLE` annotation?

geoffreygaren replied:
Yeah, it’s for the SWIFT_ESCAPABLE. Hopefully we’ll have an explicit annotation for this soon, or implicit recognition of ARC.

mwyrzykowski commented on Source/WebGPU/WebGPU/QuerySet.mm:
```diff
@@ -196,7 +196,7 @@
                 (*m_counterSampleBufferFreeRanges)[i] = newRangeSet;
                 RELEASE_ASSERT(*result / sizeof(uint64_t) < 4096);
-                return std::make_pair((*m_counterSampleBuffers)[i], *result / sizeof(uint64_t));
+                return CounterSampleBuffer { (*m_counterSampleBuffers)[i], static_cast<uint32_t>(*result / sizeof(uint64_t)) };
```
no objection but surprised the static_cast here is needed since `*result` is of type `uint32_t` and I thought uint32_t / uint64_t is still a uint32_t but also fine to leave it here

geoffreygaren replied:
C++ widens to the larger type.

mwyrzykowski replied:
and then make_pair was previously implicitly converting the uint64_t to a uint32_t? I thought that was a warning we treat as an error, I need to be more careful

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -2150,24 +2146,22 @@ extension WebGPU.CommandEncoder {
-        if m_device.ptr().enableEncoderTimestamps() || counterSampleBuffer != nil {
+        if m_device.ptr().enableEncoderTimestamps() || counterSampleBuffer.buffer != nil {
             let timestampWrites = wgpuGetComputePassDescriptorTimestampWrites(collection.span)?[0]
-            computePassDescriptor.sampleBufferAttachments[0].sampleBuffer = counterSampleBuffer != nil ? computePassDescriptor.sampleBufferAttachments[0].sampleBuffer : m_device.ptr().timestampsBuffer(m_commandBuffer, 2)
```
Line `2164` looks like it was a functional bug in the Swift version?

Should we fix that separately?  

I would have thought the following CTS case would have detected this but it passing :(: https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,encoding,beginComputePass:timestampWrites,*

geoffreygaren replied:
Yeah, that’s fair. I’ll back this out and fix separately. I guess we must be getting lucky with the pre-exiting value in the array? Not sure.

mwyrzykowski replied:
I think more likely is the CTS test is not adequately testing this and no one uses timestamp queries in compute passes in the wild

mwyrzykowski replied:
It would also be fine with me to merge it as-is since this change is overall very safe

mwyrzykowski reviewed (approved):
Thank you for fixing this, the structure is more readable than using a pair too

adetaylor commented on Source/WebGPU/WebGPU/QuerySet.h:
```diff
@@ -54,7 +54,10 @@ class Device;
+    struct CounterSampleBuffer {
+        id<MTLCounterSampleBuffer> buffer { nil }; // Safety: ARC retains this pointer
+        uint32_t offset { 0 };
+    } SWIFT_ESCAPABLE;
```
It feels to me as though `std::pair` should be labeled `SWIFT_ESCAPABLE_IF` via API notes, then this wouldn't have been needed. Could you file a radar for that?

adetaylor commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -837,27 +837,25 @@ extension WebGPU.CommandEncoder {
+                m_device.ptr().trackTimestampsBuffer(m_commandBuffer, buffer)
             } else {
-                mtlDescriptor.sampleBufferAttachments[0].sampleBuffer = counterSampleBuffer != nil ? counterSampleBuffer : m_device.ptr().timestampsBuffer(m_commandBuffer, 4)
+                mtlDescriptor.sampleBufferAttachments[0].sampleBuffer = counterSampleBuffer.buffer ?? m_device.ptr().timestampsBuffer(m_commandBuffer, 4)
```
Surely `counterSampleBuffer.buffer` would always be nil here, or we'd have gone into the other arm of the `if let`?

mwyrzykowski replied:
That's correct, and the Objective-C++ code is:
```
mtlDescriptor.sampleBufferAttachments[0].sampleBuffer = m_device->timestampsBuffer(m_commandBuffer, 4);
```

geoffreygaren committed:
[WebGPU+Swift] Eliminate 'unsafe' caused by QuerySet.CounterSampleBuffer
https://bugs.webkit.org/show_bug.cgi?id=302055
rdar://164135057

Reviewed by Mike Wyrzykowski.

Swift treats std::pair<id, int> as unsafe because

(a) there's a pointer in there

(b) Swift has no way of knowing whether the pointer's lifetime is guaranteed or
not -- because Swift models C++ copy and move constructors and copy and move
assignment operators as opaque functions.

Switched from std::pair to a custom struct so I could declare it SWIFT_ESCAPABLE.
I also took the opporutnity to give the data members more descriptive names than
'first' and 'second'.

Removed a Range(uncheckedBounds:) construction because it's unsafe and
we can use literal range syntax intead.

Changed some nil short-circuiting to use the ?? operator, for style.

Fixed up some stray ';', for style.

Canonical link: https://commits.webkit.org/302666@main
